Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,285 @@ class evector_base {
+	}
+	
+	/**
+	 *  Initialize a circular buffer
+	 *
+	 *  @param storage A storage buffer
+	 *  @return `true` on success, `false` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given storage buffer.
+	 */
+	template <size_t N>
+	bool init(T (&storage)[N]) {
+		return init(storage, N);
+	}
+	
+	/**
+	 *  Deinitialize the circular buffer
+	 */
+	void deinit() {
+		IORecursiveLockFree(lock);
+	}
+	
+	/**
+	 *  Create a circular buffer with the given capacity
+	 *
+	 *  @param size The total number of elements
+	 *  @return A non-null instance on success, `nullptr` if no memory.
+	 *  @warning The caller must invoke `CircularBuffer::destory()` to release the returned buffer.
+	 */
+	static CircularBuffer<T> *withCapacity(IOItemCount size) {
+		auto storage = IONewZero(T, size);
+		if (storage == nullptr)
+			return nullptr;
+		
+		auto instance = new CircularBuffer<T>();
+		if (instance == nullptr)
+			goto error1;
+		
+		if (!instance->init(storage, size))
+			goto error2;
+		
+		return instance;
+		
+	error2:
+		delete instance;
+		
+	error1:
+		IODelete(storage, T, size);
+		return nullptr;
+	}
+	
+	/**
+	 *  Destory the given circular buffer
+	 *
+	 *  @param buffer A non-null circular buffer returned by `CircularBuffer::withCapacity()`.
+	 */
+	static inline void destory(CircularBuffer<T> *buffer NONNULL) {
+		IOSafeDeleteNULL(buffer->storage, T, buffer->size);
+		buffer->deinit();
+		delete buffer;
+	}
+	
+	/**
+	 *  Destory the given circular buffer if it is non-null and set it to nullptr
+	 *
+	 *  @param buffer A nullable circular buffer returned by `CircularBuffer::withCapacity()`.
+	 *  @note This function mimics the macro `OSSafeReleaseNULL()`.
+	 */
+	static inline void safeDestory(CircularBuffer<T> *&buffer) {
+		if (buffer != nullptr) {
+			destory(buffer);
+			buffer = nullptr;
+		}
+	}
+	
+	/**
+	 *  Check whether the circular buffer is empty
+	 *
+	 *  @return `true` if the buffer is empty, `false` otherwise.
+	 */
+	inline bool isEmpty() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == 0) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Check whether the circular buffer is full
+	 *
+	 *  @return `true` if the buffer is full, `false` otherwise.
+	 */
+	inline bool isFull() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == size) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Get the number of elements in the circular buffer
+	 *
+	 *  @return The current number of elements in the buffer.
+	 */
+	inline IOItemCount getCount() {
+		IORecursiveLockLock(lock);
+		IOItemCount retVal = count;
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Write the given element to the circular buffer
+	 *
+	 *  @param element The element to write
+	 *  @return `true` on success, `false` if the buffer is full.
+	 */
+	inline bool write(const T &element) {
+		if (isFull())
```

Comment:
This is wrong, as there is a race condition between checking and writing. Another write may happen between isFull and locking. Please put isFull check into the recursive lock.